OrbitCaptureClient/CaptureEventProcessor.cpp

```diff
@@ -340,3 +428,254 @@ void CaptureEventProcessor::SendTracepointInfoToListenerIfNecessary(
+
+  const auto& submission_meta_info = gpu_queue_submission.meta_info();
+  const int32_t submission_thread_id = submission_meta_info.tid();
+  uint64_t submission_pre_submission_cpu_timestamp =
+      submission_meta_info.pre_submission_cpu_timestamp();
+  uint64_t submission_post_submission_cpu_timestamp =
+      submission_meta_info.post_submission_cpu_timestamp();
+
+  for (const auto& completed_marker : gpu_queue_submission.completed_markers()) {
+    CHECK(first_command_buffer != std::nullopt);
+    TimerInfo marker_timer;
+
+    // If we've recorded the submission that contains the begin marker, we'll retrieve this
+    // submission from our mappings, and set the markers begin time accordingly.
+    // Otherwise, we will use the capture start time as begin.
+    if (completed_marker.has_begin_marker()) {
+      const auto& begin_marker_info = completed_marker.begin_marker();
+      const auto& begin_marker_meta_info = begin_marker_info.meta_info();
+      const int32_t begin_marker_thread_id = begin_marker_meta_info.tid();
+      uint64_t begin_marker_post_submission_cpu_timestamp =
+          begin_marker_meta_info.post_submission_cpu_timestamp();
+      uint64_t begin_marker_pre_submission_cpu_timestamp =
+          begin_marker_meta_info.pre_submission_cpu_timestamp();
+
+      const GpuQueueSubmission* matching_begin_submission = nullptr;
+      if (submission_pre_submission_cpu_timestamp == begin_marker_pre_submission_cpu_timestamp &&
+          submission_post_submission_cpu_timestamp == begin_marker_post_submission_cpu_timestamp &&
+          submission_thread_id == begin_marker_thread_id) {
+        matching_begin_submission = &gpu_queue_submission;
+      } else {
+        matching_begin_submission = FindMatchingGpuQueueSubmission(
+            begin_marker_thread_id, begin_marker_post_submission_cpu_timestamp);
+      }
+
+      // Note, we receive submissions of a single queue in order (by CPU submission time). So if
+      // there is no matching "begin submission", the "begin" was submitted before the "end" and
+      // we lost the record of the "begin's" submission (which should not happen).
+      CHECK(matching_begin_submission != nullptr);
+
+      std::optional<GpuCommandBuffer> begin_submission_first_command_buffer =
+          FindFirstCommandBuffer(*matching_begin_submission);
+      CHECK(begin_submission_first_command_buffer.has_value());
+
+      const GpuJob* matching_begin_job = FindMatchingGpuJob(
+          begin_marker_thread_id, begin_marker_meta_info.pre_submission_cpu_timestamp(),
+          begin_marker_post_submission_cpu_timestamp);
+      CHECK(matching_begin_job != nullptr);
+
+      marker_timer.set_start(completed_marker.begin_marker().gpu_timestamp_ns() +
+                             matching_begin_job->gpu_hardware_start_time_ns() -
+                             begin_submission_first_command_buffer->begin_gpu_timestamp_ns());
```

Comment:
`matching_begin_job->gpu_hardware_start_time_ns() - begin_submission_first_command_buffer->begin_gpu_timestamp_ns()` is the difference between the cpu and the gpu time, right?